lib/email/sender.rb

```diff
@@ -514,5 +465,119 @@ def self.bounceable_reply_address?
+    #    outbound_message_id column on the Post.
+    #
+    # The RFC that covers using "Identification Fields", which are References,
+    # In-Reply-To, Message-ID, et. al. can be in the RFC link below. It's a good idea to read
+    # this beginning in the area immediately after these quotes, at least to understand
+    # the 3 main headers:
+    #
+    # > The "Message-ID:" field provides a unique message identifier that
+    # > refers to a particular version of a particular message.  The
+    # > uniqueness of the message identifier is guaranteed by the host that
+    # > generates it.
+    #
+    # > ...
+    #
+    # > The "In-Reply-To:" field may be used to identify the message (or
+    # > messages) to which the new message is a reply, while the "References:"
+    # > field may be used to identify a "thread" of conversation.
+    #
+    # https://www.rfc-editor.org/rfc/rfc5322.html#section-3.6.4
+    #
+    # It is a long read, but to understand the decision making process for this
+    # threading logic you can take a look at:
+    #
+    # https://meta.discourse.org/t/discourse-email-messages-are-incorrectly-threaded/233499
+    def add_identification_field_headers(topic, post)
+      @message.header["Message-ID"] = Email::MessageIdService.generate_or_use_existing(post)
+
+      if post.post_number > 1
+        op_message_id = Email::MessageIdService.generate_or_use_existing(topic.first_post)
+
+        ##
+        # Whenever we reply to a post directly _or_ quote a post, a PostReply
+        # record is made, with the reply_post_id referencing the newly created
+        # post, and the post_id referencing the post that was quoted or replied to.
+        referenced_posts = Post.includes(:incoming_email)
+          .joins("INNER JOIN post_replies ON post_replies.post_id = posts.id ")
+          .where("post_replies.reply_post_id = ?", post.id)
+          .order(id: :desc)
+
+        ##
+        # No referenced posts means that we are just creating a new post not
+        # referring to anything, and as such we should just fall back to using
+        # the OP.
+        if referenced_posts.empty?
+          @message.header["In-Reply-To"] = op_message_id
+          @message.header["References"] = op_message_id
+        else
+          ##
+          # When referencing _multiple_ posts then we just choose the most recent one
+          # to use for References so we have a single parent to work with, but
+          # every directly replied to post can go into In-Reply-To.
+          #
+          # We want to make sure all of the outbound_message_ids are already filled here.
+          in_reply_to_message_ids = MessageIdService.generate_or_use_existing_bulk(referenced_posts.map(&:id))
+          @message.header["In-Reply-To"] = in_reply_to_message_ids
+
+          most_recent_post = referenced_posts.first
+          most_recent_post_message_id = MessageIdService.generate_or_use_existing(most_recent_post.reload)
```

Comment:
Based on what I'm seeing now, I think we'll be able to avoid creating a bunch of ActiveRecord objects if `MessageIdService.generate_or_use_existing` accepts a `Post#id` instead and we convert the implementation to avoid relying on ActiveRecord.

Reply:
Yep that's fair, I can actually avoid this call altogether now since I have `MessageIdService.generate_or_use_existing_bulk` above and I can avoid the reload too since `in_reply_to_message_ids` will contain the most recent post. But yeah will change `generate_or_use_existing` to use SQL directly, probably add some more unit specs for that class too